### PR TITLE
Stub out concat daily file code for `padre_meddea`

### PR DESCRIPTION
**Pull request by dbarrous**
Stubs out my assumption on how to concatenate or combine partial `.fits` file into a daily `.fits` file. Please take a look and see if I'm on the right track, my first time messing around with FITs file creation outside of tweaks so feedback is welcome :smile: 

I have added specific `#TODO` comments with questions and some of the assumptions I have made. 

Also I have included a notebook for testing with files which includes the bit of testing I performed for reference. Can be removed from PR after testing.

- Also needs tests, but waiting on ensuring I am on the right track beforehand.

### CODE REVIEW

File: padre_meddea/calibration/calibration.py [hunk 1]
```diff
@@ -320,3 +320,131 @@ def read_calibration_file(calib_filename: Path):
...
+    -------
+    output_file : Path
+        Path to the concatenated daily FITS file.
+    """
+    # Check if the input files are provided
+    if not files_to_combine:
+        raise ValueError("No input files provided for concatenation.")
+
+    # Sort the files by DATE-BEG for proper time ordering
+    # TODO: Is Date-Beg the best way to sort or would it be Date-Avg/Date-ref?
```
Comment (Alrobbertz) [inline padre_meddea/calibration/calibration.py:346]: Looking at the [SOLARNET Standards](https://solarnet-metadata.readthedocs.io/en/v3.0.0/generated/parta.html#time-related-wcs-keywords) for time keywords, the `DATE-BEG` is required for all files while `DATEREF` is a special keyword for WCS coordinates. I think `DATE-BEG` is better here.  

Although, I don't think the `DATE-BEG` keyword is being added for the HK files so this may need to be a change for the HK calibration.

File: padre_meddea/calibration/calibration.py [hunk 2]
```diff
@@ -320,3 +320,131 @@ def read_calibration_file(calib_filename: Path):
...
+    # Also this could change the time of the daily file to be the first time in the list, if another time range is provided afterwards
+    date_beg = min(all_times)
+    date_end = max(all_times)
+    date_avg = date_beg + (date_end - date_beg) / 2
+
+    # TODO: What should the descriptor be for the concatenated file? I assumed `daily`
+    output_path = create_science_filename(
+        instrument,
+        time=date_beg,
+        level="l1",
```
Comment (Alrobbertz) [inline padre_meddea/calibration/calibration.py:383]: Are we still `l0` at this point?

File: padre_meddea/calibration/calibration.py [hunk 3]
```diff
@@ -320,3 +320,131 @@ def read_calibration_file(calib_filename: Path):
...
+        with fits.open(file_path) as partial_hdul:
+            log.info(f"Appending file: {file_path}")
+            # Check if the file is empty
+            if len(partial_hdul) == 0:
+                log.warning(f"File {file_path} is empty, skipping.")
+                continue
+
+            for hdu in partial_hdul[1:]:
+                # TODO: Do we need to do any checks on the HDUs here?
+                new_hdul.append(hdu.copy())
```
Comment (Alrobbertz) [inline padre_meddea/calibration/calibration.py:432]: This might vary based on the data type? We should check with Steven whether any of the data will be in HDU-0 or whether it will all be in the later HDUs.

File: padre_meddea/calibration/calibration.py [hunk 4]
```diff
@@ -320,3 +320,131 @@ def read_calibration_file(calib_filename: Path):
...
+    # TODO: Do we need to validate the concatenated file?
+    # # Before we process, validate the file with CCSDS
+    # custom_validators = [validation.validate_packet_checksums]
+    # validation_findings = validation.validate(
+    #     output_path,
+    #     valid_apids=list(padre_meddea.APID.values()),
+    #     custom_validators=custom_validators,
+    # )
+    # for finding in validation_findings:
+    #     log.warning(f"Validation Finding for File : {output_path} : {finding}")
```
Comment (Alrobbertz) [inline padre_meddea/calibration/calibration.py:443]: We shouldn't need to do CCSDS validation here. That validation code was only for CCSDS packet while converting from bin files. This concatenation process seems like its only being run for FITS files. 

We will _at some point_ want to include a SOLARNET / FITS validation routine here, but that code isn't mature yet.

**Comment by dbarrous** [comment]
@ehsteve @Alrobbertz Updated to include the concatting from Steve's notebook and included a notebook for verification that the number of rows based on Steve's notebook as well. As can be seen in the included temporary notebook there are a lot of astropy merge conflicts which I want to know what would be the best way to handle them:
```
  "WARNING: MergeConflictWarning: Cannot merge meta key 'DATE-BEG' types <class 'str'> and <class 'str'>, choosing DATE-BEG='2025-05-04T08:03:30.385' [astropy.utils.metadata.merge]\n",
      "2025-05-12 16:26:47 - astropy - WARNING: MergeConflictWarning: Cannot merge meta key 'DATE-BEG' types <class 'str'> and <class 'str'>, choosing DATE-BEG='2025-05-04T08:03:30.385'\n",
```

File: padre_meddea/calibration/calibration.py [hunk 5]
```diff
@@ -320,3 +320,162 @@ def read_calibration_file(calib_filename: Path):
...
+        Existing daily FITS file to append to.
+    outfile : Path, optional
+        Output file path. If None, will generate automatically.
+
+    Returns
+    -------
+    output_file : Path
+        Path to the concatenated daily FITS file.
+    """
+    from astropy.time import Time
```
[the remark this answers is not in the document]
Comment (ehsteve) [reply padre_meddea/calibration/calibration.py:345]: including numpy

File: padre_meddea/calibration/calibration.py [hunk 6]
```diff
@@ -320,3 +320,162 @@ def read_calibration_file(calib_filename: Path):
     # if can't read the file
 
     return None
+
+
+def concatenate_daily_fits(
```
Comment (ehsteve) [inline padre_meddea/calibration/calibration.py:325]: could you move this to fits_tools?

File: padre_meddea/calibration/calibration.py [hunk 7]
```diff
@@ -320,3 +320,162 @@ def read_calibration_file(calib_filename: Path):
...
+        Path to the concatenated daily FITS file.
+    """
+    from astropy.time import Time
+    import numpy as np
+    import logging
+
+    log = logging.getLogger(__name__)
+
+    # Validate input
+    if not files_to_combine and not existing_file:
```
Comment (ehsteve) [inline padre_meddea/calibration/calibration.py:352]: the `files_to_combine` does not have a default value so I think you would get an error without this code.

File: padre_meddea/calibration/calibration.py [hunk 8]
```diff
@@ -320,3 +320,162 @@ def read_calibration_file(calib_filename: Path):
...
+    # Calculate time range for the output file
+    all_times = []
+    for fits_file in all_files:
+        hdr = fits.getheader(fits_file)
+        for key in ("DATE-BEG", "DATE-END"):
+            if key in hdr:
+                all_times.append(Time(hdr[key]))
+
+    date_beg = Time(
+        min(all_times).datetime.replace(hour=0, minute=0, second=0, microsecond=0)
```
[the remark this answers is not in the document]
Comment (ehsteve) [reply padre_meddea/calibration/calibration.py:377]: oh i see. i don't like converting between Time objects and datetime so lets use
`Time(min(all_times).iso[0:10]))`
there may be a better way to do this but this should be safe.

**Review by ehsteve — CHANGES_REQUESTED** [review]
I tested this on my machine with a different set of test files and it all seemed to work!

Is there any way to quiet the many mergeconflict warnings?

`WARNING: MergeConflictWarning`

File: padre_meddea/calibration/calibration.py [hunk 9]
```diff
@@ -320,3 +320,162 @@ def read_calibration_file(calib_filename: Path):
     # if can't read the file
 
     return None
+
+
+def concatenate_daily_fits(
+    files_to_combine: list[Path], existing_file: Path = None, outfile: Path = None
+) -> Path:
+    """
```
Comment (ehsteve) [inline padre_meddea/calibration/calibration.py:328]: please add a Note that states that this assumes that there is no data stored in the PrimaryHDU.

**Comment by ehsteve** [comment]
@dbarrous please don't forget to add your name to the author list in the pyproject.toml file!

File: padre_meddea/io/fits_tools.py [hunk 10]
```diff
@@ -422,3 +435,745 @@ def get_prhsh(n: int = 1, a: str = "A") -> Tuple[str, str]:
...
+            hdu_list.append(new_hdu)
+        elif hdu_info["type"] == "image":
+            new_hdu = fits.ImageHDU(
+                hdu_info["data"], header=hdu_info["header"], name=hdu_info["name"]
+            )
+            hdu_list.append(new_hdu)
+
+    # Write the output file
+    hdul = fits.HDUList(hdu_list)
+    hdul.writeto(outfile, overwrite=True)
```
Comment (dbarrous) [inline padre_meddea/io/fits_tools.py:1110]: We are not using a `with` in the function which could be causing the issue in Windows. Can we add a 
`hdul.close()` right after?

**Review by ehsteve — APPROVED** [review]
Tested on my system. We might run into some edge cases and should add tests for individual functions that were added here but we can do that later.
